dmcontent/questions.py

```diff
@@ -480,6 +479,53 @@ def summary(self, service_data):
         return ListSummary(self, service_data)
 
 
+class Hierarchy(List):
+    """
+    For our purposes, a Hierarchy is like a List, except the entries
+    are potentially related to each other in a "subsumptive
+    containment hierarchy". For example, every service that is a member
+    of a child category is _by definition_ also a member of its parent
+    category as well, and we can add those parent categories as a
+    denormalization to make searching more efficient.
+    """
+
+    def _get_data(self, form_data):
+        if self.id not in form_data:
+            return {self.id: None}
+
+        values = set(form_data.getlist(self.id))
+        values.update(self.get_missing_values(values))
+
+        return {self.id: sorted(values) or None}
+
+    def summary(self, service_data):
+        return HierarchySummary(self, service_data)
+
+    def get_missing_values(self, selected_values_set):
+        """
+        Recursively retrieves un-selected parent categories of the
+        passed-in selection, as a set of 'value' strings.
+        :param selected_values_set: initially-selected categories (e.g. by the user)
+        :return: additional values that should also be selected
+        """
+        expected_values = set()
+
+        def set_expected_values(options, parents=None):
+            if parents is None:
+                parents = list()
+
+            for option in options:
+                value = option.get('value', option.get('label'))
+                if value in selected_values_set:
+                    expected_values.update(parents)
+                children = option.get('options')
+                if children:
+                    set_expected_values(children, parents + [value])
+
+        set_expected_values(self._data.get('options', []))
+        return expected_values - selected_values_set
```

Comment:
This was kinda a compromise position for @allait and me, the rationalisation being roughly:

 - really, these additional categories are a denormalization. Arguably, the frontend should just assume that they are selected - the place they actually _need_ to be selected is in the search engine
 - however, the indexer doesn't have access to the tree at the moment, so moving the denormalization there is rather hard
 - the next best place to put this might then be the API, because that would ensure a consistent view on the data. But that would require teaching the API about this as a special field, and no other fields get processing - either the data posted in is accepted, or not (i.e. we only have validation, nothing beyond that). And again, the API doesn't currently have access to the original framework data, so even if we wanted to have the additional categories added there, it's hard in the current structure.
 - that leaves us with having to do the denormalization in the frontends (i.e. in the content loader)
 - having the API validate that the denormalization _was_ applied means duplicating that content loader logic into the API, and again we don't actually have the data available that would allow us to do that anyway...
 - and therefore data _validly_ posted to the API may not have the denormalization applied (i.e. if someone posted to the API via some other method than the content loader).
 - and that's weird, but we can't do anything about it, and weird stuff shouldn't be silent, so here we are.

Ideally, I would find a way to write a pithy one-line comment somewhere that explained the above :-D